### PR TITLE
mgr/dashboard: add server side encryption to rgw/s3

### PR DESCRIPTION
**Pull request by aaSharma14**
Fixes: https://tracker.ceph.com/issues/57826
Signed-off-by: Aashish Sharma <aasharma@redhat.com>

Screenshot from 2022-09-15 16-50-02
Screenshot from 2022-09-26 18-45-48
Screenshot from 2022-09-26 18-46-08
Screenshot from 2022-09-26 18-51-07
Screenshot from 2022-09-26 18-47-18

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 1]
```diff
@@ -645,6 +645,45 @@ def set_bucket_versioning(self, bucket_name, versioning_state, mfa_delete,
                                      http_status_code=error.status_code,
                                      component='rgw')
 
+    @RestClient.api_get('/{bucket_name}?encryption')
+    def get_bucket_encryptyion(self, bucket_name, request=None):
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/services/rgw_client.py:649]:
```suggestion
    def get_bucket_encryption(self, bucket_name, request=None):
```

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 2]
```diff
@@ -645,6 +645,45 @@ def set_bucket_versioning(self, bucket_name, versioning_state, mfa_delete,
...
+            kms_master_key_element = ET.SubElement(default_encryption_element,
+                                                  'KMSMasterKeyID')
+            kms_master_key_element.text = keyId
+
+        data = ET.tostring(encryption_configuration, encoding='unicode')
+
+        try:
+            _ = request(data=data)  # type: ignore
+        except RequestException as e:
+            raise DashboardException(msg=str(e), component='rgw')
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/services/rgw_client.py:685]: We should use a decorator to wrap these exceptions (from `services/exceptions.py`) and avoid the boilerplate try-except code.

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 3]
```diff
@@ -645,6 +645,45 @@ def set_bucket_versioning(self, bucket_name, versioning_state, mfa_delete,
...
+    def get_bucket_encryptyion(self, bucket_name, request=None):
+
+        # pylint: disable=unused-argument
+        result = request()
+        if 'SSEAlgorithm' not in result:
+            result['SSEAlgorithm'] = 'No Encryption'
+        return result
+
+    @RestClient.api_put('/{bucket_name}?encryption')
+    def set_bucket_encryption(self, bucket_name, keyId, bucketKeyEnabled, sseAlgorithm, request=None):
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/services/rgw_client.py:658]: I understand that here the `mixedCase` convention has been used, but given there's already some `snake_case` over there, let's try and use the proper syntax for variable names:
```suggestion
    def set_bucket_encryption(self, bucket_name, keyId, bucketKeyEnabled, sse_algorithm, request=None):
```

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 4]
```diff
@@ -645,6 +645,45 @@ def set_bucket_versioning(self, bucket_name, versioning_state, mfa_delete,
...
+
+        bucket_key_enabled_element = ET.SubElement(rule_element,
+                                                  'BucketKeyEnabled')
+        bucket_key_enabled_element.text = bucketKeyEnabled
+
+        sse_algo_element = ET.SubElement(default_encryption_element,
+                                                  'SSEAlgorithm')
+        sse_algo_element.text = sseAlgorithm
+
+        if sseAlgorithm == 'aws:kms':
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/services/rgw_client.py:675]: Enum members are in [UPPERCASE](https://docs.python.org/3/library/enum.html):
```py
class SSEAlgorithms(Enum, str):
  NO_ENCRYPTION = 'No Encryption'
  AWS_KMS = 'aws:kms'
```
```suggestion
        if sseAlgorithm == SSEAlgorithms.AWS_KMS:
```

**Review by epuertat — COMMENTED** [review]
Good WIP! Some notes on code and styling.

**Comment by aaSharma14** [comment]
@cbodley @mattbenjamin @mdw-at-linuxbox , In this PR we have introduced the functionality to set the rgw config values for kms encryption(`rgw crypt s3 kms backend`,  `rgw crypt vault addr` etc.) through UI as mentioned here in this doc (https://docs.ceph.com/en/latest/radosgw/vault/). Further we are taking the `keyId` as user input and sending that `keyId` across the `PutBucketEncryption` API to set the default encryption for that bucket. Can you guys please confirm if this the correct implementation from dashboard side. Thanks in advance.

**Comment by cbodley** [comment]
does the dashboard allow you to enable/disable encryption policy on existing buckets, or is this just during bucket creation?

i don't really like the idea of the dashboard overriding config variables this way. what if they're already set on some/all rgws?
i'd much rather see the orchestrator deploy/configure an intregation like this. maybe the dashboard should just grey out the encryption checkbox if the rgw it's talking to doesn't have vault configured?

**Comment by aaSharma14** [comment]
> 



> does the dashboard allow you to enable/disable encryption policy on existing buckets, or is this just during bucket creation?

Thanks Casey, we can allow setting/disabling encryption on creation of a bucket and on existing buckets as well.

> i don't really like the idea of the dashboard overriding config variables this way. what if they're already set on some/all rgws? i'd much rather see the orchestrator deploy/configure an intregation like this. maybe the dashboard should just grey out the encryption checkbox if the rgw it's talking to doesn't have vault configured?

I think we can check if the config values are already set, then we can go forward with the default values and if they are not set we can allow the user to set them through dashboard. Shouldn't this be the case?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.html [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+                        </div>
+                    </div>
+                </div>
+  
+        <div class="modal-footer">
+          <cd-form-button-panel (submitActionEvent)="onSubmit()"
+                                [submitText]="actionLabels.SUBMIT"
+                                [form]="configForm"></cd-form-button-panel>
+        </div>
+        </div>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.html:249]: the `modal-footer` should be outside the `modal-body`
```suggestion
        </div>
        <div class="modal-footer">
          <cd-form-button-panel (submitActionEvent)="onSubmit()"
                                [submitText]="actionLabels.SUBMIT"
                                [form]="configForm"></cd-form-button-panel>
        </div>
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+import { RgwBucketEncryptionModel } from '../models/rgw-bucket-encryption';
+
+@Component({
+  selector: 'cd-rgw-config-modal',
+  templateUrl: './rgw-config-modal.component.html',
+  styleUrls: ['./rgw-config-modal.component.scss'],
+  providers: [RgwBucketEncryptionModel]
+})
+export class RgwConfigModalComponent implements OnInit{
+  readonly vault_address = /^((https?:\/\/)|(www.))(?:([a-zA-Z]+)|(\d+\.\d+.\d+.\d+)):\d{4}$/
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts:21]: camelCase?
```suggestion
  readonly vaultAddress = /^((https?:\/\/)|(www.))(?:([a-zA-Z]+)|(\d+\.\d+.\d+.\d+)):\d{4}$/
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+
+  kms_providers: string[];
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts:24]: same here?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+  configForm: CdFormGroup;
+
+  @Output()
+  submitAction = new EventEmitter();
+  auth_methods: string[];
+  secret_engines: string[];
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts:31]: same

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.scss [hunk 9]
```diff
@@ -0,0 +1,5 @@
+@use './src/styles/vendor/variables' as vv;
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.scss:1]: unused import?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.scss [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+
+.custom-radio {
+    padding-left: 35px;
+}
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.scss:5]: Can we instead use classes like `pl-2` or `pl-3` right next to `custom-radio` or a `div` above it?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts [hunk 11]
```diff
@@ -177,12 +212,19 @@ export class RgwBucketFormComponent extends CdForm implements OnInit {
       // Edit
       const versioning = this.getVersioningStatus();
       const mfaDelete = this.getMfaDeleteStatus();
+      console.log('inside update');
+      
+      console.log(values);
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:217]: should be removed

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+      control.updateValueAndValidity();
+    });
+  }
+
+  onSubmit() {
+    const values = this.configForm.value;
+    console.log(values);
+      this.rgwBucketService.setEncryptionConfig(values['encryptionType'], values['kms_provider'], values['auth_method']
+      , values['secret_engine'], values['secret_path'], values['namespace'], values['address'], values['token'], values['owner'], values['ssl_cert'], values['client_cert'], values['client_key']).subscribe(() => {
+        this.notificationService.show(
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts:94]: need to handle the error too.. when an error happens set the form with errors and not close it probably?

**Review by nizamial09 — COMMENTED** [review]
initial reviews

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 13]
```diff
@@ -279,6 +279,112 @@
...
+                  <span class="invalid-feedback"
+                        *ngIf="bucketForm.showError('kms_provider', frm, 'required')"
+                        i18n>This field is required.</span>
+                </div>
+              </div>
+            </div>
+
+            <div *ngIf="bucketForm.getValue('encryption_type') == 'aws:kms'">
+              <div class="form-group row">
+                <label class="cd-col-form-label"
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:369]: I see there's a validation that states this field is required?
```suggestion
                <label class="cd-col-form-label required"
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts [hunk 14]
```diff
@@ -71,20 +85,26 @@ export class RgwBucketFormComponent extends CdForm implements OnInit {
     });
     this.bucketForm = this.formBuilder.group({
       id: [null],
-      bid: [
-        null,
-        [Validators.required],
-        this.editing
-          ? []
-          : [CdValidators.bucketName(), CdValidators.bucketExistence(false, this.rgwBucketService)]
-      ],
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.ts:80]: why are we removing these validations?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html [hunk 15]
```diff
@@ -279,6 +279,112 @@
...
+                       formControlName="encryption_enabled"
+                       type="checkbox"
+                       [attr.disabled]="!kmsVaultConfig && !s3VaultConfig ? true : null">
+                <label class="custom-control-label"
+                       for="encryption_enabled"
+                       i18n>Encryption</label>
+                <cd-helper>
+                  <span i18n>Enables encryption for the objects in the bucket.
+                     To enable encryption on a bucket you need to set the configuration values for SSE-S3 or SSE-KMS.
+                     To set the configuration values <a href="#/rgw/bucket/create" (click)="openConfigModal()">Click here</a></span>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:302]: If the encryption is enabled, then I guess we don't need to show this part of the help text. We can conditionally disable this then! WDYT?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/models/rgw-bucket-encryption.ts [hunk 16]
```diff
@@ -0,0 +1,7 @@
+export class RgwBucketEncryptionModel {
+  kmsProviders = ['vault'];
+  authMethods = ['token', 'agent'];
+  secretEngines = ['kv', 'transit'];
+  sse_s3 = 'AES256';
+  sse_kms = 'aws:kms';
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/models/rgw-bucket-encryption.ts:6]: Also, do we need to assign these values in the model? Usually model will be just a `key` and then its type right?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+      .subscribe(
+        () => {
+          this.notificationService.show(
+            NotificationType.success,
+            $localize`Updated RGW Encryption Configuration values`
+          );
+          this.activeModal.close();
+          this.router.routeReuseStrategy.shouldReuseRoute = () => false;
+          this.router.onSameUrlNavigation = 'reload';
+          this.router.navigate([this.router.url]);
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts:134]: I believe the modalClose and routing should be moved to the `complete:` block of subscribe. Also, one issue I am seeing with the reload is that, it kind of resets the form. So I think its good to show a modal with a warning saying that the bucket form will be reset once you proceed. WDYT?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.html [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+        <div class="modal-footer">
+            <cd-form-button-panel (submitActionEvent)="onSubmit()"
+                                  [submitText]="actionLabels.SUBMIT"
+                                  [form]="configForm"></cd-form-button-panel>
+        </div>
+      </form>
+    </ng-container>
+  </cd-modal>
+
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.html:252]: indentation issues

**Review by nizamial09 — COMMENTED** [review]
I saw an `rgw-encryption` file in the models which is empty. that should be removed too i guess

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html (hunk 15 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html:302]: The user can change the config values whenever required.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/models/rgw-bucket-encryption.ts (hunk 16 above)
Comment (aaSharma14) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/models/rgw-bucket-encryption.ts:6]: I think(maybe am wrong) values can be initialised in a model as well, so as in future if we want to add more values to these columns, we just need to do it here, no need to change anything in the form.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts (hunk 17 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-config-modal/rgw-config-modal.component.ts:140]: At least we can create a tracker for the latter.

**Review by nizamial09 — APPROVED** [review]
I am good with as it is. Thanks @aaSharma14 for the great work! and please remember to remove the `rgw-encryption` file in the `models/` folder.

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test make check arm64

**Comment by aaSharma14** [comment]
jenkins test make check arm64

**Comment by aaSharma14** [comment]
@cbodley, I have updated the PR as per your comments, If you can please re-check that would be great. Thanks in advance.

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 19]
```diff
@@ -197,6 +197,28 @@ def _set_versioning(self, owner, daemon_name, bucket_name, versioning_state, mfa
             rgw_client.set_bucket_versioning(bucket_name, versioning_state, mfa_delete,
                                              mfa_token_serial, mfa_token_pin)
 
+    def _set_encryption(self, bid, encryption_type, keyId, daemon_name, owner):
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/controllers/rgw.py:200]:
```suggestion
    def _set_encryption(self, bid, encryption_type, key_id, daemon_name, owner):
```
don't kill me xd

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 20]
```diff
@@ -269,8 +293,10 @@ def get(self, bucket, daemon_name=None):
     def create(self, bucket, uid, zonegroup=None, placement_target=None,
                lock_enabled='false', lock_mode=None,
                lock_retention_period_days=None,
-               lock_retention_period_years=None, daemon_name=None):
+               lock_retention_period_years=None, encryption_state='false',
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/controllers/rgw.py:296]: After seeing this types of changes in tons of places I feel like we would benefit from having something like [GraphQL](https://graphql.org/learn/). 

Just a rant tho, feel free to resolve this comment xd

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 21]
```diff
@@ -330,6 +367,32 @@ def delete(self, bucket, purge_objects='true', daemon_name=None):
             'purge-objects': purge_objects
         }, json_response=False)
 
+    @RESTController.Collection(method='PUT', path='/setEncryptionConfig')
+    @allow_empty_body
+    def setEncryptionConfig(self, encryption_type=None, kms_provider=None, auth_method=None,
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/controllers/rgw.py:372]:
```suggestion
    def set_encryption_config(self, encryption_type=None, kms_provider=None, auth_method=None,
```
we should stick to snake_case in python code (there are more instances of this in these changes)

**Review by pereman2 — APPROVED** [review]

**Comment by nizamial09** [comment]
jenkins test make check

**Comment by nizamial09** [comment]
jenkins test dashboard cephadm

**Comment by nizamial09** [comment]
jenkins test dashboard

**Review by nizamial09 — APPROVED** [review]

**Pull request merged**